[Feature] Support for querying the trash used capacity

BiteTheDDDDt opened the pull request:
## Proposed changes
Support for querying the trash used capacity.
Now user can proactively scan trash directory.

## Types of changes

What types of changes does your code introduce to Doris?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] Documentation Update (if none of the other choices apply)
- [ ] Code refactor (Modify the code structure, format the code, etc...)
- [ ] Optimization. Including functional usability improvements and performance improvements.
- [ ] Dependency. Such as changes related to third-party components.
- [ ] Other.

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [X] I have created an issue on (Fix #6246) and described the bug/feature there in detail
- [X] Compiling and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [X] If these changes need document changes, I have updated the document
- [ ] Any dependent changes have been merged

BiteTheDDDDt committed:
add query 'show trash' for display backends trash used capacity

BiteTheDDDDt committed:
merge column 'BackendHost','BackendHeartBeatPort' into 'Backend' && adjust query define at 'sql_parser' && remove duplicates input backend

yangzhg commented on fe/fe-core/src/main/jflex/sql_scanner.flex:
```diff
@@ -103,6 +103,7 @@ import org.apache.doris.qe.SqlModeHelper;
         keywordMap.put("array", new Integer(SqlParserSymbols.KW_ARRAY));
         keywordMap.put("backend", new Integer(SqlParserSymbols.KW_BACKEND));
         keywordMap.put("backends", new Integer(SqlParserSymbols.KW_BACKENDS));
+        keywordMap.put("trash", new Integer(SqlParserSymbols.KW_TRASH));
```
keep the alpha order

BiteTheDDDDt replied:
done.

yangzhg commented on fe/fe-core/src/main/java/org/apache/doris/qe/ShowExecutor.java:
```diff
@@ -1690,6 +1698,46 @@ private void handleShowRoles() {
+                trashUsedCapacityB, System.currentTimeMillis() - start);
+                ok = true;
+            } catch (Exception e) {
+                LOG.warn("task exec error. backend[{}]", backend.getId(), e);
```
Just a log without further processing?

BiteTheDDDDt replied:
> 
> 
> Just a log without further processing?

now will display null when program throw error

EmmyMiao87 commented on fe/fe-core/src/main/java/org/apache/doris/analysis/ShowTrashStmt.java:
```diff
@@ -0,0 +1,94 @@
+
+    public ShowTrashStmt(List<String> backends) {
+        ImmutableMap<Long, Backend> backendsInfo = Catalog.getCurrentSystemInfo().getIdToBackend();
+        Map<String, Long> backendsID = new HashMap<String, Long>();
```
Why not use backend id directly？

BiteTheDDDDt replied:
Do you mean to directly use `'backend id'` query instead of `'backend host':'backend heart beat port'`?

EmmyMiao87 commented on fe/fe-core/src/main/java/org/apache/doris/analysis/ShowTrashStmt.java:
```diff
@@ -0,0 +1,94 @@
+
+public class ShowTrashStmt extends ShowStmt {
+    public static final ImmutableList<String> TITLE_NAMES = new ImmutableList.Builder<String>().add("BackendId")
+            .add("Backend").add("TrashUsedCapacity").build();
```
Would it be better to add the trash information to show proc as well?

BiteTheDDDDt replied:
> 
> 
> Would it be better to add the trash information to show proc as well?

now support 'show proc '/trash''

EmmyMiao87 commented on fe/fe-core/src/main/cup/sql_parser.cup:
```diff
@@ -2554,6 +2554,14 @@ show_param ::=
     {:
         RESULT = new ShowBackendsStmt();
     :}
+    | KW_TRASH KW_ON LPAREN string_list:backends RPAREN
```
Please add document of show trash stmt

BiteTheDDDDt replied:
> 
> 
> Please add document of show trash stmt

done.

yangzhg commented on be/src/service/backend_service.cpp:
```diff
@@ -210,6 +210,20 @@ void BackendService::get_tablet_stat(TTabletStatResult& result) {
+    std::vector<DataDirInfo> data_dir_infos;
+    StorageEngine::instance()->get_all_data_dir_info(&data_dir_infos, false /*do not update */);
+
+    for (auto& root_path_info : data_dir_infos) {
```
```suggestion
    for (const auto& root_path_info : data_dir_infos) {
```

yangzhg commented on be/src/olap/storage_engine.cpp:
```diff
@@ -377,6 +377,20 @@ OLAPStatus StorageEngine::get_all_data_dir_info(vector<DataDirInfo>* data_dir_in
+        return std::filesystem::file_size(file_path);
+    }
+    int64_t sum_size = 0;
+    for (auto& it : std::filesystem::directory_iterator(file_path)) {
```
```suggestion
    for (const auto& it : std::filesystem::directory_iterator(file_path)) {
```

BiteTheDDDDt committed:
1. change sql_scannner.flex to keep the alpha order.
2. display null when connect fail to client
3. trash information to show proc, and support show proc '/trash'
4. add document of show trash stmt

BiteTheDDDDt committed:
fix bug about document of show trash stmt

BiteTheDDDDt committed:
Update be/src/olap/storage_engine.cpp

Co-authored-by: Zhengguo Yang <780531911@qq.com>

BiteTheDDDDt committed:
Update be/src/service/backend_service.cpp

Co-authored-by: Zhengguo Yang <780531911@qq.com>

BiteTheDDDDt committed:
remove useless TITLENAME define

yangzhg commented on fe/fe-core/src/main/java/org/apache/doris/common/proc/TrashProcNode.java:
```diff
@@ -0,0 +1,110 @@
+                Pair<Double, String> trashUsedCapacity = DebugUtil.getByteUint(trashUsedCapacityB);
+                backendInfo.add(DebugUtil.DECIMAL_FORMAT_SCALE_3.format(trashUsedCapacity.first) + " " + trashUsedCapacity.second);
+            } else {
+                backendInfo.add("null");
```
not literal `null`

BiteTheDDDDt replied:
fixed

BiteTheDDDDt committed:
change literal null to empty string

BiteTheDDDDt committed:
add support for 'SHOW PROC /trash/backendId'

BiteTheDDDDt committed:
add limit about query format like 'trash/backendId/backendId'

BiteTheDDDDt committed:
1. Removed the syntax for querying multiple backends.
2. Support for querying the specific disk information of a single backend.
3. Updated related interfaces and documents.

yangzhg reviewed (dismissed):
LGTM

morningman commented on fe/fe-core/src/main/java/org/apache/doris/analysis/ShowTrashDiskStmt.java:
```diff
@@ -0,0 +1,79 @@
+        if (ConnectContext.get().getSessionVariable().getForwardToMaster()) {
+            return RedirectStatus.FORWARD_NO_SYNC;
+        } else {
+            return RedirectStatus.NO_FORWARD;
```
No need to forward

morningman commented on fe/fe-core/src/main/java/org/apache/doris/analysis/ShowTrashStmt.java:
```diff
@@ -0,0 +1,77 @@
+        if (ConnectContext.get().getSessionVariable().getForwardToMaster()) {
+            return RedirectStatus.FORWARD_NO_SYNC;
+        } else {
+            return RedirectStatus.NO_FORWARD;
```
No need to forward

BiteTheDDDDt replied:
I fix it at c3f7c7e

morningman reviewed (commented):
ShowTrashDiskStmt and ShowTrashStmt can be unified

BiteTheDDDDt committed:
change Redirect to 'NO_FORWARD' at ShowTrashStmt/ShowTrashDiskStmt

morningman reviewed (approved):
LGTM
